File: frontend/app/src/main/java/com/h1oo7/musicapp/ui/fragment/SearchFragment.java
package com.h1oo7.musicapp.ui.fragment;

import android.os.Bundle;
import android.os.Handler;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.h1oo7.musicapp.R;
import com.h1oo7.musicapp.model.Song;
import com.h1oo7.musicapp.model.response.SongResponse;
import com.h1oo7.musicapp.network.ApiService;
import com.h1oo7.musicapp.network.RetrofitClient;
import com.h1oo7.musicapp.ui.adapter.SongAdapter;
import com.google.android.material.textfield.TextInputEditText;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class SearchFragment extends Fragment {

    private TextInputEditText etSearch;
    private RecyclerView recyclerResults;
    private TextView tvNoResults;

    private SongAdapter adapter;

    private Handler searchHandler = new Handler();
    private Runnable searchRunnable;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {

        View view = inflater.inflate(R.layout.fragment_search, container, false);

        etSearch = view.findViewById(R.id.et_search);
        recyclerResults = view.findViewById(R.id.recycler_search_results);
        tvNoResults = view.findViewById(R.id.tv_no_results);

        recyclerResults.setLayoutManager(new LinearLayoutManager(requireContext()));
        adapter = new SongAdapter();
        recyclerResults.setAdapter(adapter);

        setupSearchListener();

        return view;
    }

    private void setupSearchListener() {
        searchRunnable = () -> {
            String keyword = etSearch.getText().toString().trim();
            if (keyword.isEmpty()) {
                adapter.setSongs(null);
                tvNoResults.setVisibility(View.GONE);
                return;
            }
            searchSongs(keyword);
        };

        etSearch.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                searchHandler.removeCallbacks(searchRunnable);
                searchHandler.postDelayed(searchRunnable, 400); // debounce 400ms
            }

            @Override
            public void afterTextChanged(Editable s) {}
        });
    }

    private void searchSongs(String keyword) {
        ApiService api = RetrofitClient.getClient().create(ApiService.class);

        api.getAllSongs(1, 50, keyword).enqueue(new Callback<SongResponse>() {
            @Override
            public void onResponse(Call<SongResponse> call, Response<SongResponse> response) {
                if (response.isSuccessful() && response.body() != null) {

                    List<Song> results = response.body().getSongs();
                    adapter.setSongs(results);

                    if (results == null || results.isEmpty()) {
                        tvNoResults.setVisibility(View.VISIBLE);
                    } else {
                        tvNoResults.setVisibility(View.GONE);
                    }
                }
            }

            @Override
            public void onFailure(Call<SongResponse> call, Throwable t) {
                tvNoResults.setVisibility(View.VISIBLE);
            }
        });
    }
}

File: forntend/app/src/main/java/com/example/musique/ui/fragments/SearchFragment.kt
package com.example.musique.ui.fragments

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.musique.data.network.RetrofitClient
import com.example.musique.data.repository.SongRepository
import com.example.musique.databinding.FragmentSearchBinding
import com.example.musique.player.MusicPlayerManager
import com.example.musique.ui.adapter.SongAdapter
import com.example.musique.ui.viewmodel.SongViewModel
import com.example.musique.ui.viewmodel.ViewModelFactory
import com.example.musique.utils.PreferenceManager
import com.example.musique.utils.Resource
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class SearchFragment : Fragment() {
    
    private var _binding: FragmentSearchBinding? = null
    private val binding get() = _binding!!
    
    private lateinit var viewModel: SongViewModel
    private lateinit var songAdapter: SongAdapter
    private lateinit var musicPlayerManager: MusicPlayerManager
    
    private var searchJob: Job? = null
    
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentSearchBinding.inflate(inflater, container, false)
        return binding.root
    }
    
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        
        setupViewModel()
        setupRecyclerView()
        setupObservers()
        setupSearchListener()
        
        musicPlayerManager = MusicPlayerManager.getInstance(requireContext())
        
        viewModel.getSongs(null, 1, 20)
    }
    
    private fun setupViewModel() {
        val preferenceManager = PreferenceManager(requireContext())
        val apiService = RetrofitClient.getApiService(preferenceManager)
        val repository = SongRepository(apiService)
        val factory = ViewModelFactory(songRepository = repository)
        viewModel = ViewModelProvider(this, factory)[SongViewModel::class.java]
    }
    
    private fun setupRecyclerView() {
        songAdapter = SongAdapter(
            onSongClick = { song ->
                viewModel.setCurrentSong(song)
                musicPlayerManager.playSong(song)
                viewModel.incrementPlayCount(song._id)
            },
            onPlayClick = { song ->
                viewModel.setCurrentSong(song)
                musicPlayerManager.playSong(song)
                viewModel.incrementPlayCount(song._id)
            }
        )
        
        binding.rvSongs.apply {
            layoutManager = LinearLayoutManager(requireContext())
            adapter = songAdapter
        }
    }
    
    private fun setupObservers() {
        viewModel.songsState.observe(viewLifecycleOwner) { resource ->
            when (resource) {
                is Resource.Loading -> {
                    binding.progressBar.visibility = View.VISIBLE
                    binding.tvError.visibility = View.GONE
                }
                is Resource.Success -> {
                    binding.progressBar.visibility = View.GONE
                    binding.tvError.visibility = View.GONE
                    songAdapter.submitList(resource.data?.songs)
                }
                is Resource.Error -> {
                    binding.progressBar.visibility = View.GONE
                    binding.tvError.visibility = View.VISIBLE
                    binding.tvError.text = resource.message
                }
            }
        }
    }
    
    private fun setupSearchListener() {
        binding.etSearch.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                searchJob?.cancel()
                searchJob = CoroutineScope(Dispatchers.Main).launch {
                    delay(500)
                    val keyword = s?.toString()?.trim()
                    viewModel.getSongs(keyword?.ifEmpty { null }, 1, 20)
                }
            }
            
            override fun afterTextChanged(s: Editable?) {}
        })
    }
    
    override fun onDestroyView() {
        super.onDestroyView()
        searchJob?.cancel()
        _binding = null
    }
}
